MRTKOculusConfig script update

chetu3319 opened the pull request:
MRTKOculusConfig:
- Updating `instance` variable which holds the clone of the config file.
- Updating `CustomHandMaterial` variable which on modification fire an event.

OculusQuestInputManager:
- Adding a Listener to update hand material on update.

Commit:
MRTKOculusConfig script update

MRTKOculusConfig:
- Updating `instance` variable  which holds the clone of the config file.
- Updating `CustomHandMaterial` variable which on modification fire an event.

OculusQuestInputManager:
- Adding Listener to update hand material on update.

chetu3319 commented:
Upon material update, there is a visible transition glitch. 
I assume there to be a better way to update the material on hands instead of calling `RemoveAllHAndDevices()` function.

provencher commented on .vscode/settings.json:
```diff
@@ -0,0 +1,56 @@
+﻿{
```
Please remove this file from the PR. IDE config files should be ignored.

chetu3319 replied:
Removed the File.

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Config/MRTKOculusConfig.cs:
```diff
@@ -106,10 +108,25 @@ public static MRTKOculusConfig Instance
+            set
+            {
+                customHandMaterial = value;
+                if (onCustomHandMaterialUpdate != null)
```
you can do a ternary null check for events like this.
onCustomHandMaterialUpdate?.Invoke() I believe

chetu3319 replied:
Thank you for the suggestion. 
Updated the requested change

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Config/MRTKOculusConfig.cs:
```diff
@@ -106,10 +108,25 @@ public static MRTKOculusConfig Instance
+        {
+            get
+            {
+                return customHandMaterial;
```
get =>

chetu3319 replied:
Made changes according to the coding style.

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Config/MRTKOculusConfig.cs:
```diff
@@ -106,10 +108,25 @@ public static MRTKOculusConfig Instance
         [Tooltip("Custom hand material to use for hand tracking hand mesh.")]
         private Material customHandMaterial = null;
 
+        public System.Action onCustomHandMaterialUpdate;
```
public variables have uppercase to start -> OnCustomHandMaterialUpdate

chetu3319 replied:
Made the change.

provencher commented on Packages/manifest.json:
```diff
@@ -1,14 +1,26 @@
+    "com.unity.analytics": "3.3.2",
+    "com.unity.collab-proxy": "1.2.16",
+    "com.unity.ext.nunit": "1.0.0",
+    "com.unity.ide.rider": "1.1.0",
```
lots of unnecessary package changes - please uncheck these from your commit.

chetu3319 replied:
Reverted.

provencher commented on ProjectSettings/ProjectVersion.txt:
```diff
@@ -1 +1,2 @@
-m_EditorVersion: 2018.4.22f1
+m_EditorVersion: 2019.2.19f1
```
Careful working with 2019 when making PRs, as this repo is setup with 2018.4 for now, at least until 2019.3 is the official version recommended for MRTK.

chetu3319 replied:
Reverted.

provencher commented on ProjectSettings/ProjectSettings.asset:
```diff
@@ -49,7 +49,7 @@ PlayerSettings:
   defaultScreenHeight: 768
```
Please remove this file from your PR

provencher commented on ProjectSettings/XRSettings.asset:
```diff
@@ -0,0 +1,10 @@
+{
```
please remove this file from your pr

chetu3319 replied:
Deleted.

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Input/OculusQuestInputManager.cs:
```diff
@@ -307,6 +309,21 @@ protected void UpdateHand(OVRHand ovrHand, OVRSkeleton ovrSkeleton, OVRMeshRende
+
+            leftHandMaterial = new Material(MRTKOculusConfig.Instance.CustomHandMaterial);
+
+            RemoveAllHandDevices();
```
Instead of calling remove all hand device, you should look at how the material is used in the OculusQuestHand. 
You should create a function there that updates the material on the renderer, and call that here and in the constructor for OculusQuestHand. Destroying the hands is a bit overkill.

chetu3319 replied:
Created a public function in `OculusQuestHand.cs` script for updating material to the renderer.

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Config/MRTKOculusConfig.cs:
```diff
@@ -43,7 +43,9 @@ public static MRTKOculusConfig Instance
                 if (instance == null)
                 {
-                    instance = Resources.Load<MRTKOculusConfig>("MRTK-OculusConfig");
+                    // Creating a configurable instance so that making any runtime changes won't affect the original. 
```
Thanks for adding a comment.

provencher reviewed (changes requested):
A few changes to make, but you're on the right track!
Big thing to note is to be careful when working in a different unity version than the repo is made with, and especially avoid modifying settings and the json manifest.

Commit:
Removing IDE specific setting file

Removing IDE specific setting file

Commit:
Reverting Setting files

Revreting following setting files to the original:
- `manifest.json` : reverting package file to the original one.
- `ProjectSettings.asset`
- `ProjectVersion.txt` : reverting to 2018.4.22f1
- `XRSettings.asset` : Removing XRSettings package file.

Commit:
Code change for smooth update of Hand Material

- Implementing request changes in PR adopting the coding format.
- Creating new function for updating Hand Material in `OculusQuestHand.cs` script.

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Input/OculusQuestInputManager.cs:
```diff
@@ -307,6 +309,32 @@ protected void UpdateHand(OVRHand ovrHand, OVRSkeleton ovrSkeleton, OVRMeshRende
 
+        public void UpdateHandMaterial()
+        {
+            if (rightHandMaterial)
```
Please add curly braces

provencher replied:
Also an explicit null check

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Input/OculusQuestInputManager.cs:
```diff
@@ -307,6 +309,32 @@ protected void UpdateHand(OVRHand ovrHand, OVRSkeleton ovrSkeleton, OVRMeshRende
+
+            rightHandMaterial = new Material(MRTKOculusConfig.Instance.CustomHandMaterial);
+
+            if (leftHandMaterial)
```
curly braces + explicit null check

chetu3319 replied:
Done

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Input/OculusQuestInputManager.cs:
```diff
@@ -158,6 +159,7 @@ public override void Disable()
 
             RemoveAllControllerDevices();
             RemoveAllHandDevices();
+            MRTKOculusConfig.Instance.OnCustomHandMaterialUpdate -= UpdateHandMaterial;
```
Id remove the listener at the top to be safe

chetu3319 replied:
Are you pointing me to move Line 162 up to Line 159?  Or just eliminating the unsubscription to the event. 
Can you describe it?

provencher replied:
Sorry I hadn't seen your reply.

My idea is just to move the unsub part a little higher in the function call, to make sure the removing calls don't somehow trip the callback.

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Config/MRTKOculusConfig.cs:
```diff
@@ -106,10 +108,20 @@ public static MRTKOculusConfig Instance
         [Tooltip("Custom hand material to use for hand tracking hand mesh.")]
         private Material customHandMaterial = null;
 
+        public System.Action OnCustomHandMaterialUpdate;
```
Please add a new line between OnCustomHandMaterialUpdate and the custom hand material property

provencher replied:
Also please add a summary block for the action, as it is public

chetu3319 replied:
Done

provencher commented on Assets/MixedRealityToolkit.ThirdParty/MRTK-Quest/Scripts/Input/Controllers/OculusQuestHand.cs:
```diff
@@ -84,6 +84,13 @@ public OculusQuestHand(TrackingState trackingState, Handedness controllerHandedn
+        {
+
             if (!MRTKOculusConfig.Instance.UseCustomHandMaterial) return;
 
```
Please update this.handMaterial when updating the material

chetu3319 replied:
Currently, `this.handMaterial` is a read-only private variable. I'm unsure if there is any butterfly effect of changing it. 

This function `UpdateHandMaterial()` updates the material to the current instance of the hand by directly tapping into the renderer. When the tracking is lost and regained, a new instance of quest hand will be created and with that `this.handMaterial` will store the updated material. 

I can see a drawback, where if any code which will try to read the `handMaterial` will receive old material for that particular instance.

provencher replied:
You definitely need to update this.handMaterial, or the pinch affordance won’t work until the new hand is made.

It’s read only because it wasn’t replaced at runtime before, but it’s not that big a deal to change it.

chetu3319 replied:
Thank you for your guidance. Made necessary changes.

provencher reviewed (changes requested):
Few more small changes - almost there!

Commit:
Clean up

- Added Summary tags to public variables
- Code clean up

provencher reviewed (approved):
I do think this adds a bit of complexity, but I think the value add is nice, and the implementation is clean.
